File: 2021/julia/3.jl
function get_most_common(diagnostics, bit_position)
    total = 0

    # loop over each diagnostic and sum the digits
    for i in 1:length(diagnostics)
        total += parse(Int, diagnostics[i][bit_position])
    end;

    # divide by the total number of diagnostics
    total /= length(diagnostics)

    return total < 0.5 ? 0 : 1
end;

function get_power(diagnostics)
    # count how many digits in each binary number
    n_digit = length(diagnostics[1])

    most_common = ""
    least_common = ""

    # loop over each digit
    for i in 1:n_digit
        most = get_most_common(diagnostics, i)
        # this gives most common and least common
        most_common *= string(most)
        least_common *= string(1 - most)
    end;

    # convert numbers to decimals and multiply
    power = parse(Int, most_common, base=2) * parse(Int, least_common, base=2)

    # part 1 done!
    return power
end;


function get_gas_rating(diagnostics, bit_position, least)
    # find the most common bit at this bit position
    most_common = get_most_common(diagnostics, bit_position)
    criterion = least ? 1 - most_common : most_common

    # create a reduced list of diagnostics that match the criterion
    reduced_diagnostics = String[]
    for i in 1:length(diagnostics)
        if string(diagnostics[i][bit_position]) == string(criterion)
            push!(reduced_diagnostics, diagnostics[i])
        end;
    end;

    # if only one left then return it
    if length(reduced_diagnostics) == 1
        return parse(Int, reduced_diagnostics[1], base=2)
    # otherwise recurse
    else
        return get_gas_rating(reduced_diagnostics, bit_position + 1, least)
    end;
end;


function get_life_support(diagnostics)
    return get_gas_rating(diagnostics, 1, false) * get_gas_rating(diagnostics, 1, true)
end;


function main()
    diagnostics = open("../inputs/3.txt", "r") do input
        diagnostics = [chomp(line) for line in eachline(input)]
    end;

    println("PART ONE: ", get_power(diagnostics))
    println("PART TWO: ", get_life_support(diagnostics))
end;

main()

File: 2021/python/3.py

def binary_to_decimal(binary):
    decimal = 0
    for i in range(len(binary)):
        decimal += int(binary[i]) * 2**(len(binary) - 1 - i)
    return decimal


def main():
    # read in diagnostics
    with open("../inputs/3.txt") as input:
        diagnostics = [line.replace("\n", "") for line in input]

    # count how many digits in each binary number
    n_digit = len(diagnostics[0])

    most_common = ""
    least_common = ""

    # loop over each digit
    for i in range(n_digit):
        total = 0

        # loop over each diagnostic and sum the digits
        for j in range(len(diagnostics)):
            total += int(diagnostics[j][i])

        # divide by the total number of diagnostics
        total /= len(diagnostics)

        # this gives most common and least common
        most_common += str(0) if total < 0.5 else str(1)
        least_common += str(1) if total < 0.5 else str(0)

    # convert numbers to decimals and multiply
    power = binary_to_decimal(most_common) * binary_to_decimal(least_common)

    # part 1 done!
    print(power)


if __name__ == "__main__":
    main()
